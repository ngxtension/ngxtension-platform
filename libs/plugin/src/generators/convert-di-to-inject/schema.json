{
	"$schema": "http://json-schema.org/schema",
	"$id": "",
	"title": "",
	"type": "object",
	"properties": {
		"project": {
			"type": "string",
			"aliases": ["name", "projectName"],
			"description": "Project for which to convert from constructor DI to inject function."
		},
		"path": {
			"type": "string",
			"description": "Path to the component/directive/pipe you want to convert to use the inject function"
		},
		"includeReadonlyByDefault": {
			"type": "boolean",
			"aliases": ["includeReadonly", "includeReadonlyByDefault"],
			"description": "Whether to include readonly for injections by default. Default is false."
		},
		"useEsprivateFieldNotation": {
			"type": "boolean",
			"aliases": [
				"useESPrivateFieldNotation",
				"useEsprivateFieldNotation",
				"useEsPrivateFieldNotation"
			],
			"description": "Whether to replace TS private modifier with ES private field notation. Default is false."
		}
	}
}
