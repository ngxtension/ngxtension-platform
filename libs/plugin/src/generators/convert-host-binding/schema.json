{
	"$schema": "http://json-schema.org/schema",
	"$id": "ConvertHostBinding",
	"title": "",
	"type": "object",
	"properties": {
		"project": {
			"type": "string",
			"aliases": ["name", "projectName"],
			"description": "Project for which to convert from host binding to host property."
		},
		"path": {
			"type": "string",
			"description": "Path to the component/directive you want to convert to use the host property."
		}
	}
}
