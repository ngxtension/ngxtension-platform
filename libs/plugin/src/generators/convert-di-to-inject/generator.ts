import {
	formatFiles,
	getProjects,
	logger,
	readJson,
	readProjectConfiguration,
	Tree,
	visitNotIgnoredFiles,
} from '@nx/devkit';
import { readFileSync } from 'node:fs';
import { exit } from 'node:process';
import { Node, Scope, VariableDeclarationKind } from 'ts-morph';
import { ContentsStore } from '../shared-utils/contents-store';
import { ConvertDiToInjectGeneratorSchema } from './schema';

function trackContents(
	tree: Tree,
	contentsStore: ContentsStore,
	fullPath: string,
) {
	if (fullPath.endsWith('.ts')) {
		const fileContent =
			tree.read(fullPath, 'utf8') || readFileSync(fullPath, 'utf8');

		if (
			!fileContent.includes('@Component') &&
			!fileContent.includes('constructor(') &&
			!fileContent.includes('@Pipe') &&
			!fileContent.includes('@Injectable') &&
			!fileContent.includes('@Directive')
		) {
			return;
		}

		if (fileContent.includes('constructor(')) {
			contentsStore.track(fullPath, fileContent);
		}
	}
}

function tokenIsTypeString(token: any) {
	// example: 'my-service' or "my-service"
	return token.includes("'") || token.includes('"');
}

export async function convertDiToInjectGenerator(
	tree: Tree,
	options: ConvertDiToInjectGeneratorSchema,
) {
	const contentsStore = new ContentsStore();
	const packageJson = readJson(tree, 'package.json');
	const angularCorePackage = packageJson['dependencies']['@angular/core'];

	if (!angularCorePackage) {
		logger.error(`[ngxtension] No @angular/core detected`);
		return exit(1);
	}

	const { path, project } = options;

	if (path && project) {
		logger.error(
			`[ngxtension] Cannot pass both "path" and "project" to convertDiToInjectGenerator`,
		);
		return exit(1);
	}

	if (path) {
		if (!tree.exists(path)) {
			logger.error(`[ngxtension] "${path}" does not exist`);
			return exit(1);
		}

		trackContents(tree, contentsStore, path);
	} else if (project) {
		try {
			const projectConfiguration = readProjectConfiguration(tree, project);

			if (!projectConfiguration) {
				throw `"${project}" project not found`;
			}

			visitNotIgnoredFiles(tree, projectConfiguration.root, (path) => {
				trackContents(tree, contentsStore, path);
			});
		} catch (err) {
			logger.error(`[ngxtension] ${err}`);
			return;
		}
	} else {
		const projects = getProjects(tree);
		for (const project of projects.values()) {
			visitNotIgnoredFiles(tree, project.root, (path) => {
				trackContents(tree, contentsStore, path);
			});
		}
	}

	for (const { path: sourcePath } of contentsStore.collection) {
		const sourceFile = contentsStore.project.getSourceFile(sourcePath)!;

		const hasInjectImport = sourceFile.getImportDeclaration(
			(importDecl) =>
				importDecl.getModuleSpecifierValue() === '@angular/core' &&
				importDecl
					.getNamedImports()
					.some((namedImport) => namedImport.getName() === 'inject'),
		);

		const classes = sourceFile.getClasses();

		for (const targetClass of classes) {
			const applicableDecorator = targetClass.getDecorator((decoratorDecl) => {
				return ['Component', 'Directive', 'Pipe', 'Injectable'].includes(
					decoratorDecl.getName(),
				);
			});
			if (!applicableDecorator) continue;

			const convertedDeps = new Set<string>();
			const convertedPrivateDeps = new Set<string>();

			let includeHostAttributeToken = false;

			targetClass.getConstructors().forEach((constructor) => {
				constructor.getParameters().forEach((param, index) => {
					const { name, type, decorators, scope, isReadonly } =
						param.getStructure();

					let shouldUseType = false;
					let toBeInjected = type; // default to type
					let tokenToBeInjectedIsString = false;
					let isAttributeToken = false;
					const flags = [];

					if (decorators.length > 0) {
						decorators.forEach((decorator) => {
							if (decorator.name === 'Inject') {
								toBeInjected = decorator.arguments[0]; // use the argument of the @Inject decorator
								if (toBeInjected !== type) {
									shouldUseType = true;
								}

								if (tokenIsTypeString(toBeInjected)) {
									tokenToBeInjectedIsString = true;
								}
							}

							if (decorator.name === 'Attribute') {
								// ex: @Attribute('type') type: string
								toBeInjected = decorator.arguments[0];
								isAttributeToken = true;
								includeHostAttributeToken = true;
								shouldUseType = true;
							}

							if (decorator.name === 'Optional') {
								flags.push('optional');
							}
							if (decorator.name === 'Self') {
								flags.push('self');
							}
							if (decorator.name === 'SkipSelf') {
								flags.push('skipSelf');
							}
							if (decorator.name === 'Host') {
								flags.push('host');
							}
						});
					}

					// if type is (ElementRef or TemplateRef) or should use type, add it as inject generic

					let injection = 'inject';

					const typeHasGenerics = type?.toString().includes('<') ?? false;

					if (type && (shouldUseType || typeHasGenerics)) {
						injection += `<${type}>`;
					}

					let initializer = '';

					if (isAttributeToken) {
						// inject(new HostAttributeToken('some-attr'));
						initializer = `${injection}(new HostAttributeToken(${toBeInjected})${flags.length > 0 ? `, { ${flags.map((flag) => flag + ': true').join(', ')} }` : ''})`;
					} else {
						initializer = `${injection}(${toBeInjected}${tokenToBeInjectedIsString ? ' as any /* TODO(inject-migration): Please check if the type is correct */' : ''}${flags.length > 0 ? `, { ${flags.map((flag) => flag + ': true').join(', ')} }` : ''})`;
					}

					if (!scope) {
						// create the variable inside the constructor instead of creating it as a class property
						constructor.insertVariableStatement(0, {
							declarationKind: VariableDeclarationKind.Const,
							declarations: [{ name, initializer }],
						});

						// check if the service was used inside the constructor with 'this.' prefix
						// if so, remove the 'this.' prefix and use the service variable directly
						constructor.getStatements().forEach((statement) => {
							if (!Node.isExpressionStatement(statement)) {
								return;
							}

							const callExpression = statement.getExpression();
							if (!Node.isCallExpression(callExpression)) {
								return;
							}

							const memberExpression = callExpression.getExpression();
							if (!Node.isMemberExpression(memberExpression)) {
								return;
							}

							const text = memberExpression.getText();
							const [thisKeyword, serviceName] = text.split('.');
							if (thisKeyword === 'this' && serviceName === name.toString()) {
								const newExpression = text.replace(
									`this.${name.toString()}`,
									name.toString(),
								);
								memberExpression.replaceWithText(newExpression);
							}
						});
					} else {
						const hasPrivateScope = scope === Scope.Private;
						const propertyName =
							hasPrivateScope && options.useEsprivateFieldNotation
								? `#${name}`
								: name;

						if (hasPrivateScope) {
							convertedPrivateDeps.add(name);
						}

						targetClass.insertProperty(index, {
							name: propertyName,
							initializer,
							scope:
								hasPrivateScope && options.useEsprivateFieldNotation
									? null
									: scope,
							isReadonly:
								isReadonly || options.includeReadonlyByDefault || false,
							leadingTrivia: '  ',
						});

						const updates = [] as (() => void)[];
						param
							.findReferences()
							.flatMap((ref) => ref.getReferences())
							.filter((ref) => !ref.isDefinition())
							.forEach((ref) => {
								const node = ref.getNode();
								const parent = node.getParent();

								const text = parent.getText();
								if (text.includes(`this.${name}`)) {
									return;
								}
								updates.push(() => {
									parent.replaceWithText(
										text.replace(name.toString(), `this.${propertyName}`),
									);
								});
							});

						updates.forEach((update) => update());
					}

					convertedDeps.add(name);
				});

				if (convertedDeps.size > 0 && !hasInjectImport) {
					const namedImports = ['inject'];

					if (includeHostAttributeToken) {
						namedImports.push('HostAttributeToken');
					}

					const angularCoreImports = sourceFile.getImportDeclaration(
						(importDecl) => {
							return importDecl.getModuleSpecifierValue() === '@angular/core';
						},
					);
					if (angularCoreImports) {
						angularCoreImports.addNamedImports(namedImports);
					} else {
						sourceFile.insertImportDeclaration(0, {
							namedImports,
							moduleSpecifier: '@angular/core',
							leadingTrivia: '  ',
						});
					}
				}

				constructor.getParameters().forEach((param) => {
					if (convertedDeps.has(param.getName())) {
						param.remove();
					}
				});

				if (
					constructor.getParameters().length === 0 &&
					constructor.getBodyText().trim() === ''
				) {
					constructor.remove();
				}

				for (const decorator of [
					'Inject',
					'Attribute',
					'Optional',
					'Self',
					'SkipSelf',
					'Host',
				]) {
					// if @${Decorator} is not used anymore, remove the import
					const hasDecoratorUsage = sourceFile
						.getFullText()
						.includes(`@${decorator}`);
					if (!hasDecoratorUsage) {
						const foundImport = sourceFile.getImportDeclaration(
							(importDecl) =>
								importDecl.getModuleSpecifierValue() === '@angular/core' &&
								importDecl
									.getNamedImports()
									.some((namedImport) => namedImport.getName() === decorator),
						);

						if (foundImport) {
							const classToRemove = foundImport
								.getNamedImports()
								.find((namedImport) => namedImport.getName() === decorator);
							classToRemove.remove();
						}
					}
				}
			});

			if (options.useEsprivateFieldNotation) {
				Array.from(convertedPrivateDeps).forEach((convertedDepsName) => {
					const startIndex = targetClass.getProperties().length;
					const tempAddedProperty = targetClass.insertProperty(startIndex, {
						name: convertedDepsName,
						leadingTrivia: '  ',
					});

					tempAddedProperty
						.findReferences()
						.flatMap((ref) => ref.getReferences())
						.filter((ref) => !ref.isDefinition())
						.forEach((ref) => {
							const node = ref.getNode();
							const parent = node.getParent();

							const text = parent.getText();
							if (
								text.includes(`this.#${convertedDepsName}`) ||
								text.includes(`#${convertedDepsName}`)
							) {
								return;
							}

							parent.replaceWithText(
								text.replace(convertedDepsName, `#${convertedDepsName}`),
							);
						});

					tempAddedProperty.remove();
				});
			}
		}

		tree.write(sourcePath, sourceFile.getFullText());
	}

	await formatFiles(tree);

	logger.info(
		`
[ngxtension] Conversion completed. Please check the content and run your formatter as needed.
`,
	);
}

export default convertDiToInjectGenerator;
