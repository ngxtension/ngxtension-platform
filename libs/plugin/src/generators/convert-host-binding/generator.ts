import {
	getProjects,
	logger,
	readJson,
	readProjectConfiguration,
	Tree,
	visitNotIgnoredFiles,
} from '@nx/devkit';
import { readFileSync } from 'node:fs';
import { exit } from 'node:process';
import { SyntaxKind } from 'ts-morph';
import { ContentsStore } from '../shared-utils/contents-store';
import { ConvertHostBindingGeneratorSchema } from './schema';

const CLASS_DECORATORS = ['Component', 'Directive'];

function trackContents(
	tree: Tree,
	contentsStore: ContentsStore,
	fullPath: string,
) {
	if (fullPath.endsWith('.ts')) {
		const fileContent =
			tree.read(fullPath, 'utf8') || readFileSync(fullPath, 'utf8');

		const decorators = CLASS_DECORATORS.map(
			(decoratorName) => `@${decoratorName}`,
		);

		if (!decorators.some((decorator) => fileContent.includes(decorator))) {
			logger.log(
				`[ngxtension] "${fullPath}" is not a Component nor a Directive`,
			);
			return;
		}
		if (
			fileContent.includes('@HostBinding') ||
			fileContent.includes('@HostListener')
		) {
			contentsStore.track(fullPath, fileContent);
		}
	}
}

export async function convertHostBindingGenerator(
	tree: Tree,
	options: ConvertHostBindingGeneratorSchema,
) {
	const contentsStore = new ContentsStore();
	const packageJson = readJson(tree, 'package.json');
	const angularCorePackage = packageJson['dependencies']['@angular/core'];

	if (!angularCorePackage) {
		logger.error(`[ngxtension] No @angular/core detected`);
		return exit(1);
	}

	const { path, project } = options;

	if (path && project) {
		logger.error(
			`[ngxtension] Cannot pass both "path" and "project" to convertHostBinding`,
		);
		return exit(1);
	}

	/** Track contents in the store */
	if (path) {
		if (!tree.exists(path)) {
			logger.error(`[ngxtension] "${path}" does not exist`);
			return exit(1);
		}

		trackContents(tree, contentsStore, path);
	} else if (project) {
		try {
			const projectConfiguration = readProjectConfiguration(tree, project);

			if (!projectConfiguration) {
				throw `"${project}" project not found`;
			}

			visitNotIgnoredFiles(tree, projectConfiguration.root, (path) => {
				trackContents(tree, contentsStore, path);
			});
		} catch (err) {
			logger.error(`[ngxtension] ${err}`);
			return;
		}
	} else {
		const projects = getProjects(tree);
		for (const project of projects.values()) {
			visitNotIgnoredFiles(tree, project.root, (path) => {
				trackContents(tree, contentsStore, path);
			});
		}
	}
	/** end Track contents in the store */

	for (const { path: sourcePath } of contentsStore.collection) {
		const sourceFile = contentsStore.project.getSourceFile(sourcePath);
		addHostProperty(sourceFile);
		removeHostBindingOrHostListenerImport(sourceFile);
		tree.write(sourcePath, sourceFile.getFullText());
	}
}

const removeHostBindingOrHostListenerImport = (sourceFile) => {
	const packageDir = '@angular/core';
	const foundImports = sourceFile
		.getImportDeclaration(
			(importDeclaration) =>
				importDeclaration.getModuleSpecifierValue() === packageDir,
		)
		.getNamedImports()
		.filter(
			(namedImport) =>
				namedImport.getName() === 'HostBinding' ||
				namedImport.getName() === 'HostListener',
		);

	if (foundImports.length) {
		foundImports.forEach((foundImport) => foundImport.remove());
	}
};

const addHostProperty = (sourceFile) => {
	sourceFile.getClasses().forEach((klass) => {
		const decorator = klass.getDecorator((decoratorDecl) =>
			CLASS_DECORATORS.includes(decoratorDecl.getName()),
		);
		if (!decorator) {
			return;
		}
		const hostBindings: Record<string, string> = {};
		const hostListeners: Record<string, string> = {};

		const processHostDecorators = (elements, hostBindings, hostListeners) => {
			elements.forEach((element) => {
				const hostBindingDecorator = element.getDecorator('HostBinding');
				if (hostBindingDecorator) {
					const argument = hostBindingDecorator.getArguments()[0];
					const binding = argument
						? getHostBindingKey(argument.getText())
						: element.getName();

					hostBindings[binding] = isMethodDeclaration(element)
						? `'${element.getName()}()'`
						: `'${element.getName()}'`;
					/** Remove the HostBinding decorator */
					hostBindingDecorator.remove();
				}

				const hostListenerDecorator = element.getDecorator('HostListener');
				if (hostListenerDecorator) {
					const [event, ...args] = hostListenerDecorator.getArguments();
					const binding = getHostListenerKey(event.getText());
					hostListeners[binding] =
						args.length && element.getParameters().length
							? `'${element.getName()}(${args.map((arg) => arg.getText().replace(/^\['|'\]$/g, '')).join(', ')})'`
							: `'${element.getName()}()'`;

					/** Remove the HostListener decorator */
					hostListenerDecorator.remove();
				}
			});
		};

		const processProperties = () => {
			processHostDecorators(klass.getProperties(), hostBindings, hostListeners);
		};

		const processGetters = () => {
			processHostDecorators(
				klass.getGetAccessors(),
				hostBindings,
				hostListeners,
			);
		};

		const processMethods = () => {
			processHostDecorators(klass.getMethods(), hostBindings, hostListeners);
		};

		processProperties();
		processGetters();
		processMethods();

		const resolveHostProperties = (hostProperties) => {
			if (Object.keys(hostProperties).length > 0) {
				const hostProperty = decorator.getArguments()[0].getProperty('host');
				if (hostProperty) {
					const existingHostBindings = hostProperty.getInitializerIfKindOrThrow(
						SyntaxKind.ObjectLiteralExpression,
					);
					Object.entries(hostProperties).forEach(([key, value]) => {
						existingHostBindings.addPropertyAssignment({
							name: key,
							initializer: value,
						});
					});
				} else {
					decorator.getArguments()[0].addPropertyAssignment({
						name: 'host',
						initializer: `{ ${Object.entries(hostProperties)
							.map(([key, value]) => `${key}: ${value}`)
							.join(', ')} }`,
					});
				}
			}
		};

		const resolveHostBindings = (hostBindings) => {
			resolveHostProperties(hostBindings);
		};

		const resolveHostListeners = (hostListeners) => {
			resolveHostProperties(hostListeners);
		};

		resolveHostBindings(hostBindings);
		resolveHostListeners(hostListeners);
	});
};

const getHostBindingKey = (hostBinding: string) => {
	return `'[${removeQuotes(hostBinding)}]'`;
};

const getHostListenerKey = (hostListener: string) => {
	return `'(${removeQuotes(hostListener)})'`;
};

function removeQuotes(str: string) {
	return str.replace(/['"]/g, '');
}

function isMethodDeclaration(node) {
	return node.getKindName() === 'MethodDeclaration';
}

export default convertHostBindingGenerator;
