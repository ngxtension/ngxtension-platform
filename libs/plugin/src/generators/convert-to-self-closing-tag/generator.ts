import {
	HtmlParser,
	ParseTreeResult,
	RecursiveVisitor,
	Text,
	visitAll,
} from '@angular-eslint/bundled-angular-compiler';
import { Element as Element_2 } from '@angular/compiler';
import {
	Tree,
	formatFiles,
	getProjects,
	joinPathFragments,
	logger,
	readJson,
	readProjectConfiguration,
	visitNotIgnoredFiles,
} from '@nx/devkit';
import { readFileSync } from 'node:fs';
import { dirname } from 'node:path';
import { exit } from 'node:process';
import { Node, SyntaxKind } from 'ts-morph';
import { ContentsStore } from '../shared-utils/contents-store';
import { ConvertToSelfClosingTagGeneratorSchema } from './schema';

function trackContents(
	tree: Tree,
	contentsStore: ContentsStore,
	fullPath: string,
) {
	if (fullPath.endsWith('.ts')) {
		const fileContent =
			tree.read(fullPath, 'utf8') || readFileSync(fullPath, 'utf8');
		if (!fileContent.includes('@Component')) return;
		if (
			fileContent.includes('template') ||
			fileContent.includes('templateUrl')
		) {
			contentsStore.track(fullPath, fileContent);
		}
	}
}

export async function convertToSelfClosingTagGenerator(
	tree: Tree,
	options: ConvertToSelfClosingTagGeneratorSchema,
) {
	const contentsStore = new ContentsStore();
	const packageJson = readJson(tree, 'package.json');
	const angularCorePackage =
		packageJson['dependencies']['@angular/core'] ||
		packageJson['devDependencies']['@angular/core'];

	if (!angularCorePackage) {
		logger.error(`[ngxtension] No @angular/core detected`);
		return exit(1);
	}

	const { path, project } = options;

	if (path && project) {
		logger.error(
			`[ngxtension] Cannot pass both "path" and "project" to convertToSelfClosingTagGenerator`,
		);
		return exit(1);
	}

	if (path) {
		if (!tree.exists(path)) {
			logger.error(`[ngxtension] "${path}" does not exist`);
			return exit(1);
		}

		trackContents(tree, contentsStore, path);
	} else if (project) {
		try {
			const projectConfiguration = readProjectConfiguration(tree, project);

			if (!projectConfiguration) {
				throw `"${project}" project not found`;
			}

			visitNotIgnoredFiles(tree, projectConfiguration.root, (path) => {
				trackContents(tree, contentsStore, path);
			});
		} catch (err) {
			logger.error(`[ngxtension] ${err}`);
			return;
		}
	} else {
		const projects = getProjects(tree);
		for (const project of projects.values()) {
			visitNotIgnoredFiles(tree, project.root, (path) => {
				trackContents(tree, contentsStore, path);
			});
		}
	}

	for (const { path: sourcePath } of contentsStore.collection) {
		const sourceFile = contentsStore.project.getSourceFile(sourcePath)!;

		const classes = sourceFile.getClasses();

		for (const targetClass of classes) {
			const applicableDecorator = targetClass.getDecorator((decoratorDecl) => {
				return ['Component'].includes(decoratorDecl.getName());
			});
			if (!applicableDecorator) continue;

			// process decorator metadata references
			const decoratorArg = applicableDecorator.getArguments()[0];
			if (Node.isObjectLiteralExpression(decoratorArg)) {
				decoratorArg
					.getChildrenOfKind(SyntaxKind.PropertyAssignment)
					.forEach((property) => {
						const decoratorPropertyName = property.getName();

						if (decoratorPropertyName === 'template') {
							let originalText = property.getFullText();
							originalText = migrateComponentToSelfClosingTags(originalText);

							if (originalText !== property.getFullText()) {
								property.replaceWithText(originalText.trimStart());
							}
						} else if (decoratorPropertyName === 'templateUrl') {
							const dir = dirname(sourcePath);
							const templatePath = joinPathFragments(
								dir,
								property
									.getInitializer()
									.getText()
									.slice(1, property.getInitializer().getText().length - 1),
							);
							let templateText = tree.exists(templatePath)
								? tree.read(templatePath, 'utf8')
								: '';

							if (templateText) {
								templateText = migrateComponentToSelfClosingTags(templateText);
								tree.write(templatePath, templateText);
							}
						}
					});
			}
		}

		tree.write(sourcePath, sourceFile.getFullText());
	}

	if (contentsStore.withTransforms.size) {
		logger.info(
			`
[ngxtension] The following classes have had some Inputs with "transform" converted. Please double check the type arguments on the "transform" Inputs
`,
		);
		contentsStore.withTransforms.forEach((className) => {
			logger.info(`- ${className}`);
		});
	}

	await formatFiles(tree);

	logger.info(
		`
[ngxtension] Conversion completed. Please check the content and run your formatter as needed.
`,
	);
}

export function migrateComponentToSelfClosingTags(template: string): string {
	const parsedTemplate: ParseTreeResult = new HtmlParser().parse(
		template,
		'template.html',
		{ tokenizeBlocks: true },
	);

	const visitor = new ElementCollector();
	visitAll(visitor, parsedTemplate.rootNodes);

	let changedOffset = 0;
	visitor.elements.forEach((element) => {
		const { start, end, tagName } = element;

		const currentLength = template.length;
		const templatePart = template.slice(
			start + changedOffset,
			end + changedOffset,
		);

		function replaceWithSelfClosingTag(html, tagName) {
			const pattern = new RegExp(
				`<\\s*${tagName}\\s*([^>]*?(?:"[^"]*"|'[^']*'|[^'">])*)\\s*>([\\s\\S]*?)<\\s*/\\s*${tagName}\\s*>`,
				'gi',
			);
			const replacement = `<${tagName} $1 />`;
			return html.replace(pattern, replacement);
		}

		const convertedTemplate = replaceWithSelfClosingTag(templatePart, tagName);

		// if the template has changed, replace the original template with the new one
		if (convertedTemplate.length !== templatePart.length) {
			template = replaceTemplate(
				template,
				convertedTemplate,
				start,
				end,
				changedOffset,
			);

			changedOffset += template.length - currentLength;
		}
	});

	return template;
}

class ElementCollector extends RecursiveVisitor {
	readonly elements: ElementToMigrate[] = [];

	constructor() {
		super();
	}

	override visitElement(element: Element_2, context: any) {
		if (element.children.length) {
			if (element.children.length === 1) {
				const child = element.children[0];

				if (child instanceof Text) {
					if (child.value.trim() === '' || child.value.trim() === '\n') {
						if (element.name.includes('-')) {
							this.elements.push({
								tagName: element.name,
								start: element.sourceSpan.start.offset,
								end: element.sourceSpan.end.offset,
							});
						}
					}
				}
			}

			return super.visitElement(element, context);
		}

		if (element.name.includes('-')) {
			this.elements.push({
				tagName: element.name,
				start: element.sourceSpan.start.offset,
				end: element.sourceSpan.end.offset,
			});
		}
		return super.visitElement(element, context);
	}
}

/**
 * Replace the value in the template with the new value based on the start and end position + offset
 */
function replaceTemplate(
	template: string,
	replaceValue: string,
	start: number,
	end: number,
	offset: number,
) {
	return (
		template.slice(0, start + offset) +
		replaceValue +
		template.slice(end + offset)
	);
}

export interface ElementToMigrate {
	tagName: string;
	start: number;
	end: number;
}

export default convertToSelfClosingTagGenerator;
