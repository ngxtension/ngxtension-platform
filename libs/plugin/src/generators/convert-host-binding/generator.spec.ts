import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';

import { convertHostBindingGenerator } from './generator';
import { ConvertHostBindingGeneratorSchema } from './schema';

const filesMap = {
	componentWithDuplicateHostBinding: `
  import { Component, HostBinding } from '@angular/core';

  @Component({
    selector: 'my-duplicated-component',
    template: 'My duplicated component',
  })
  export class MyDuplicatedComponent {
    @HostBinding('class.active') isActive = true;
    @HostBinding('class.active') isActive2 = true;
  `,
	componentWithComplexHostListener: `
  import { Component, HostListener } from '@angular/core';
  @Component({
    selector: 'my-complex-component',
    template: 'My complex component',
  })
  export class MyComplexComponent {
    @HostListener('document:keydown.escape', ['$event']) escapeKeydownHandler(event: KeyboardEvent): void {
    }

    @HostListener('mousedown', ['$event.target'])
    handleMousedown(target: any): void {
    }
    @HostListener('click', ['$event'])
    handleEventNoArgs(): void {
    }
  }
  `,
	componentWithNoHostBinding: `
  import { Component } from '@angular/core';

  @Component({
    selector: 'my-empty-component',
    template: 'My empty component',
  })
  export class MyEmptyComponent {
  }
  `,
	componentWithHostBinding: `
	import { Component, HostBinding } from '@angular/core';

  @Component({
    selector: 'my-component',
    template: 'My component',
  })
  export class MyComponent {
    @HostBinding('class.active') isActive = true;

    @HostBinding('attr.aria-disabled') get isDisabled() {
      return true;
    }

    @HostBinding('tabIndex') getTabIndex() {
      return this.isDisabled ? -1 : 0;
    }
  }
  `,
	directiveWithHostBinding: `
	import { Directive, HostBinding, HostListener } from '@angular/core';

  @Directive({
    selector: '[myDirective]'
  })
  export class MyDirective {
    @HostBinding('class.active') isActive = true;

    @HostListener('keydown', ['$event'])
    updateValue(event: KeyboardEvent) {
    }
  }
  `,

	componentWithHostProperty: `
  import { Component } from '@angular/core';

    @Component({
      selector: 'my-component',
      template: 'My component',
      host: {
        '[class.active]': 'isActive',
      }
    })
    export class MyComponent {
      isActive = true;

      @HostBinding('attr.aria-disabled') get isDisabled() {
        return true;
      }
    }
  `,
};

describe('convert-host-binding generator', () => {
	let tree: Tree;
	const options: ConvertHostBindingGeneratorSchema = {
		path: 'libs/my-file.ts',
	};

	function setup(file: keyof typeof filesMap) {
		tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
		tree.write('package.json', `{"dependencies": {"@angular/core": "17.3.0"}}`);
		tree.write(`libs/my-file.ts`, filesMap[file]);

		return () => {
			return [tree.read('libs/my-file.ts', 'utf8'), filesMap[file]];
		};
	}

	beforeEach(() => {
		tree = createTreeWithEmptyWorkspace();
	});

	it('should convert properly for component', async () => {
		const readContent = setup('componentWithHostBinding');
		await convertHostBindingGenerator(tree, options);

		const [updated] = readContent();

		expect(updated).toMatchSnapshot();
	});

	it('should convert properly for directive', async () => {
		const readContent = setup('directiveWithHostBinding');
		await convertHostBindingGenerator(tree, options);

		const [updated] = readContent();

		expect(updated).toMatchSnapshot();
	});

	it('should convert properly for component with host property', async () => {
		const readContent = setup('componentWithHostProperty');
		await convertHostBindingGenerator(tree, options);

		const [updated] = readContent();

		expect(updated).toMatchSnapshot();
	});

	it('should convert properly for component with no host binding', async () => {
		const readContent = setup('componentWithNoHostBinding');
		await convertHostBindingGenerator(tree, options);

		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should convert properly for component with complex host listener', async () => {
		const readContent = setup('componentWithComplexHostListener');
		await convertHostBindingGenerator(tree, options);

		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should convert properly for component with duplicate host binding', async () => {
		const readContent = setup('componentWithDuplicateHostBinding');
		await convertHostBindingGenerator(tree, options);

		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});
});
