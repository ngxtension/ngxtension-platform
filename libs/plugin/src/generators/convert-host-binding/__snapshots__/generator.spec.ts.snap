// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert-host-binding generator should convert properly for component 1`] = `
"
	import { Component } from '@angular/core';

  @Component({
    selector: 'my-component',
    template: 'My component',
      host: { '[class.active]': 'isActive', '[attr.aria-disabled]': 'isDisabled', '[tabIndex]': 'getTabIndex()' }
})
  export class MyComponent {
    isActive = true;

    get isDisabled() {
      return true;
    }

    getTabIndex() {
      return this.isDisabled ? -1 : 0;
    }
  }
  "
`;

exports[`convert-host-binding generator should convert properly for component with complex host listener 1`] = `
"
  import { Component } from '@angular/core';
  @Component({
    selector: 'my-complex-component',
    template: 'My complex component',
      host: { '(document:keydown.escape)': 'escapeKeydownHandler($event)', '(mousedown)': 'handleMousedown($event.target)', '(click)': 'handleEventNoArgs()' }
})
  export class MyComplexComponent {
    escapeKeydownHandler(event: KeyboardEvent): void {
    }

    
    handleMousedown(target: any): void {
    }
    
    handleEventNoArgs(): void {
    }
  }
  "
`;

exports[`convert-host-binding generator should convert properly for component with duplicate host binding 1`] = `
"
  import { Component } from '@angular/core';

  @Component({
    selector: 'my-duplicated-component',
    template: 'My duplicated component',
      host: { '[class.active]': 'isActive2' }
})
  export class MyDuplicatedComponent {
    isActive = true;
    isActive2 = true;
  "
`;

exports[`convert-host-binding generator should convert properly for component with host property 1`] = `
"
  import { Component } from '@angular/core';

    @Component({
      selector: 'my-component',
      template: 'My component',
      host: {
        '[class.active]': 'isActive',
          '[attr.aria-disabled]': 'isDisabled'
    }
    })
    export class MyComponent {
      isActive = true;

      get isDisabled() {
        return true;
      }
    }
  "
`;

exports[`convert-host-binding generator should convert properly for component with no host binding 1`] = `
"
  import { Component } from '@angular/core';

  @Component({
    selector: 'my-empty-component',
    template: 'My empty component',
  })
  export class MyEmptyComponent {
  }
  "
`;

exports[`convert-host-binding generator should convert properly for directive 1`] = `
"
	import { Directive } from '@angular/core';

  @Directive({
    selector: '[myDirective]',
      host: { '[class.active]': 'isActive',
          '(keydown)': 'updateValue($event)'
    }
})
  export class MyDirective {
    isActive = true;

    
    updateValue(event: KeyboardEvent) {
    }
  }
  "
`;
